README.md
# Auth::Middleware

Allow any app to be protected behind a login-wall using QS ID

## Usage

The middleware is a standard Rack middleware use it as such.

This makes the app only accesible when logged in:

```ruby
# config.ru

require 'auth/middleware'

authed_app = Rack::Builder.new do
  use Auth::Middleware, 'my-qs-app-id', 'my-qs-app-secret', 'qs_auth_cookie_name' do |auth_tools|
    auth_tools.require_login!
  end
  run MyOriginalApp
end

run authed_app
```

You can also allow only admins to access your app:

```ruby
# config.ru

require 'auth/middleware'

authed_app = Rack::Builder.new do
  use Auth::Middleware, 'my-qs-app-id', 'my-qs-app-secret', 'qs_auth_cookie_name' do |auth_tools|
    auth_tools.require_admin!
  end
  run MyOriginalApp
end

run authed_app
```

CHANGELOG.md
# 0.0.2 / 2013-08-01

* Adds endpoint to log out

# 0.0.1 / 2013-07-31

* The beginning

spec/spec_helper.rb
ENV['RACK_ENV'] ||= 'test'
ENV['QS_AUTH_BACKEND_URL'] = 'http://auth-backend.dev'

Bundler.require

require 'minitest/autorun'

require 'auth/middleware'
require 'qs-test-harness'

class MiddlewareInjector
  def initialize(app)
    @app = app
    self.class.instances << self
  end

  def self.instances
    @instances ||= []
  end

  def self.use(middleware, *args, &blck)
    @middleware = middleware
    @middleware_args = args
    @middleware_blck = blck
  end

  def self.middleware
    @middleware
  end

  def self.middleware_args
    @middleware_args
  end

  def self.middleware_blck
    @middleware_blck
  end

  def self.reset!
    instances.each(&:reset!)
  end

  def reset!
    @middleware = nil
  end

  def call(env)
    setup! unless setup?
    @middleware.call(env)
  end

  protected
  def setup!
    raise "No middleware set. Can't use the MiddlewareInjector" unless self.class.middleware

    if self.class.middleware_blck
      @middleware = self.class.middleware.new(@app, *(self.class.middleware_args || []), &self.class.middleware_blck)
    else
      @middleware = self.class.middleware.new(@app, *(self.class.middleware_args || []))
    end
  end

  def setup?
    !!@middleware
  end
end

class AuthClientAugmenter
  def <<(client)
    @clients ||= []
    @clients << client
    process!
  end

  def app=(app)
    @app = app
    process!
  end

  def process!
    return unless @app
    while client = @clients.shift
      client.instance_variable_set('@adapter', Service::Client::Adapter::Faraday.new(adapter: [:rack, @app]))
    end
  end
end

QS_AUTH_CLIENT_AUGMENTER = AuthClientAugmenter.new

require 'auth-client'
module Auth
  class Client
    alias raw_initialize initialize
    def initialize(*args)
      raw_initialize(*args)
      QS_AUTH_CLIENT_AUGMENTER << self
    end
  end
end

class SampleApp
  def self.new
    Rack::Builder.new do
      use Rack::Session::Cookie, key: 'qs_auth_middleware_test', secret: '1234567890'
      use MiddlewareInjector

      run lambda {|e| [200, {'Content-Type' => 'text/plain'}, ["Secret - #{e['qs_auth_tools'].token_owner['uuid']}"]]}
    end
  end
end

class OauthInjectorApp
  def self.app=(app)
    @app = app
  end

  def self.app
    @app
  end

  def self.call(env)
    request = Rack::Request.new(env)

    app.call(env)
  end
end

require 'oauth2'
module OAuth2
  class Client
    # The Faraday connection object
    def connection
      @__connection ||= begin
        conn = Faraday.new(site, options[:connection_opts])
        conn.build do |b|
          options[:connection_build].call(b) if options[:connection_build]
        end
        conn.request :url_encoded
        conn.adapter :rack, OauthInjectorApp
        conn
      end
    end
  end
end

Qs::Test::Harness.setup! do
  provide Qs::Test::Harness::Provider::Datastore
  provide Qs::Test::Harness::Provider::Graph
  provide Qs::Test::Harness::Provider::Auth

  test SampleApp
end

OauthInjectorApp.app = Qs::Test::Harness.harness.provider(:auth).app

QS_AUTH_CLIENT_AUGMENTER.app = Qs::Test::Harness.harness.provider(:auth).app

oauth_app = Qs::Test::Harness.harness.entity_factory.create(:app, redirect_url: "http://example.com/auth/auth_backend/callback")

MIDDLEWARE_OAUTH_APP = oauth_app

spec/middleware_spec.rb
require_relative './spec_helper'

require 'rack/client'
require 'uri'
require 'cgi'

describe Auth::Middleware do
  before do
    @harness = Qs::Test::Harness.harness
    @client = @harness.client
  end

  def handle_oauth_procedure(initial_response, user)
    cookie = user.logged_in_cookie

    app_cookie = initial_response.headers['Set-Cookie']

    response = @client.get initial_response.headers['Location'], headers: {'Referer' => "http://example.com/afterloginthing", 'Cookie' => app_cookie}
    response.status.must_equal 302
    response.headers['Location'].must_match /\/oauth\/authorize/
    app_cookie = response.headers['Set-Cookie']

    url = response.headers['Location']
    state = url.match(/(&|\?)state=(.*)(&.*)?$/).captures[1]
    state.wont_be_nil
    state.wont_be_empty

    auth_client = @harness.provider(:auth).client
    response = auth_client.get url, headers: {'Cookie' => cookie, 'Referer' => 'http://example.com/'}
    response.status.must_equal 200
    response.body.index('"/oauth/allow"').wont_be_nil
    response.body.index(state).wont_be_nil
    cookie = response.headers['Set-Cookie']

    params = {
      response_type: 'code',
      client_id: MIDDLEWARE_OAUTH_APP.id,
      redirect_uri: MIDDLEWARE_OAUTH_APP.redirect_url,
      state: state,
      allow: '1'
    }
    body = URI.encode_www_form(params)
    response = auth_client.post "/oauth/allow", headers: {'Cookie' => cookie}, body: body
    response.status.must_equal 302

    response = @client.get response.headers['Location'], headers: {'Cookie' => app_cookie}
    response.status.must_equal 301
    response.headers['Location'].must_equal "http://example.com/afterloginthing"
    app_cookie = response.headers['Set-Cookie']

    [app_cookie, @client.get(response.headers['Location'], headers: {'Cookie' => app_cookie})]
  end

  describe "login required" do
    before do
      MiddlewareInjector.reset!
      MiddlewareInjector.use Auth::Middleware, MIDDLEWARE_OAUTH_APP.id, MIDDLEWARE_OAUTH_APP.secret, 'qs_auth_middleware_test' do |auth_tools|
        auth_tools.require_login!
      end
    end

    it "shows you the secret page after login" do
      response = @client.get 'http://example.com/'
      response.status.must_equal 302
      response.body.must_be_empty

      user = @harness.entity_factory.create(:user)
      cookie, response = handle_oauth_procedure(response, user)

      response.status.must_equal 200
      response.body.must_equal "Secret - #{user.uuid}"
    end

    it "has a logout endpoint" do
      response = @client.get 'http://example.com/'
      user = @harness.entity_factory.create(:user)
      cookie, response = handle_oauth_procedure(response, user)
      response.status.must_equal 200
      response.body.must_equal "Secret - #{user.uuid}"

      response = @client.get 'http://example.com/', headers: {'Cookie' => cookie}
      response.status.must_equal 200
      response.body.must_equal "Secret - #{user.uuid}"

      response = @client.get 'http://example.com/auth/auth_backend/logout', headers: {'Cookie' => cookie}
      response.status.must_equal 302
      response.headers['Location'].must_equal "#{ENV['QS_AUTH_BACKEND_URL']}/signout?redirect_uri=#{CGI.escape('http://example.com/')}"

      cookie_expiration = response.headers['Set-Cookie'].match(/expires=(.*)(;.*)?$/i).captures.first
      date, time, junk = cookie_expiration.split(', ').last.split(' ')
      day, month, year = date.split('-')
      hour, minute, second = time.split(':')
      cookie_expiration = Time.new(year, month, day, hour.to_i, minute.to_i, second.to_i)

      cookie_expiration < Time.now
    end
  end

  describe "admin privileges required" do
    before do
      MiddlewareInjector.reset!
      MiddlewareInjector.use Auth::Middleware, MIDDLEWARE_OAUTH_APP.id, MIDDLEWARE_OAUTH_APP.secret, 'qs_auth_middleware_test' do |auth_tools|
        auth_tools.require_admin!
      end
    end

    it "redirects you back to login when logged in but not as an admin" do
      response = @client.get 'http://example.com/'
      response.status.must_equal 302
      response.body.must_be_empty

      user = @harness.entity_factory.create(:user, :admin => false)
      cookie, response = handle_oauth_procedure(response, user)

      response.status.must_equal 200
      response.body.must_equal "No access without admin privileges."

      response = @client.get 'http://example.com/', headers: {'Cookie' => cookie}
      response.status.must_equal 200
      response.body.must_equal "No access without admin privileges."
    end

    it "shows you the secret page when logged in as an admin" do
      response = @client.get 'http://example.com/'
      response.status.must_equal 302
      response.body.must_be_empty

      user = @harness.entity_factory.create(:user, :admin => true)
      cookie, response = handle_oauth_procedure(response, user)
      response.status.must_equal 200
      response.body.must_equal "Secret - #{user.uuid}"
    end
  end
end

lib/auth/middleware.rb
require "auth/middleware/version"
require "auth/middleware/auth_tools"
require "auth/middleware/qs_strategy"

require 'auth-client'
require 'uri'
require 'cgi'

module Auth
  class Middleware
    attr_reader :options, :cookie_name

    def initialize(app, app_id, app_secret, cookie_name, &blck)
      handled_app = lambda do |env|
        case env['PATH_INFO']
        when "/auth/auth_backend/callback"
          response = Rack::Response.new('', 301, 'Location' => env['omniauth.origin'] || '/')
          response.set_cookie(@cookie_name, value: JSON.dump(env['omniauth.auth']), path: '/')
          response
        when "/auth/auth_backend/logout"
          request = Rack::Request.new(env)
          come_back_url = URI.join(request.url, '/').to_s

          response = Rack::Response.new('', 302, 'Location' => "#{ENV['QS_AUTH_BACKEND_URL']}/signout?redirect_uri=#{CGI.escape(come_back_url)}")
          response.set_cookie(@cookie_name, value: '', path: '/', expires: Time.new(1970))
          response
        else
          app.call(env)
        end
      end

      @omniauth_strategy = QsStrategy.new(handled_app, app_id, app_secret)
      @cookie_name = cookie_name
      @options = options

      @auth_client = Auth::Client.new(ENV['QS_AUTH_BACKEND_URL'])

      @auth_block = blck
    end

    def call(env)
      if request_should_be_handled?(env)
        auth_tools = env['qs_auth_tools'] = AuthTools.new(self, env)
        if auth_cookie = auth_tools.cookie
          return auth_tools.force_logout! unless auth_tools.token_owner = @auth_client.token_owner(auth_cookie['info']['token'])
        end

        if @auth_block
          response = catch(:response) do
            @auth_block.call(auth_tools)
          end
          return response if response && response.kind_of?(Rack::Response)
        end
      end

      @omniauth_strategy.call(env)
    end

    protected
    def request_should_be_handled?(env)
      !env['PATH_INFO'].match(/^\/auth\/auth_backend/)
    end
  end
end


lib/auth/middleware/auth_tools.rb
require 'json'
require 'cgi'
require 'uri'

module Auth
  class Middleware
    class AuthTools
      attr_accessor :token_owner
      attr_reader :request

      def initialize(middleware, env)
        @middleware = middleware
        @env = env
        @request = Rack::Request.new(env)
      end

      def cookie
        raw_cookie = @request.cookies()[@middleware.cookie_name]
        return nil unless raw_cookie
        @cookie ||= JSON.parse(raw_cookie)
      end

      def require_login!
        redirect!('/auth/auth_backend', true) unless token_owner
      end

      def require_admin!
        require_login!

        render_message!("No access without admin privileges.") unless admin?
      end

      def force_logout!
        redirect_url = "#{ENV['QS_AUTH_BACKEND_URL']}/signout?redirect_uri=#{CGI.escape(request.url)}"
        response = Rack::Response.new('', 301, 'Location' => redirect_url)
        response.set_cookie(@cookie_name, :value => '', :path => "/", :expires => Time.new(1970).utc)
        response
      end

      def admin?
        token_owner && token_owner['admin'] == true
      end

      protected
      def redirect!(url, same_host = false)
        response = Rack::Response.new('', 302)
        if same_host
          uri = URI.parse(request.url)
          uri.path = url
          uri.query = nil
          uri.fragment = nil
          url = uri.to_s
        end
        response.redirect(url)
        throw(:response, response)
      end

      def render_message!(message)
        headers = {
            'Content-Type' => 'text/plain',
            'Content-Length' => message.bytesize
        }
        headers['Set-Cookie'] = @env['HTTP_COOKIE'] if @env['HTTP_COOKIE']
        Rack::Response.new(message, 200, headers)
      end
    end
  end
end